Mathlib/Order/Hom/Bounded.lean

```diff
@@ -0,0 +1,829 @@
+  [BoundedOrder β] extends OrderHom α β where
+  /-- The function preserves the top element. The preferred spelling is `map_top`. -/
+  map_top' : toFun ⊤ = ⊤
+  /-- The function preserves the bottom element. The preferred spelling is `map_bot`. -/
+  map_bot' : toFun ⊥ = ⊥
+#align bounded_order_hom BoundedOrderHom
+
+section
+
+/-- `top_hom_class F α β` states that `F` is a type of `⊤`-preserving morphisms.
+
+You should extend this class when you extend `top_hom`. -/
+class TopHomClass (F : Type _) (α β : outParam <| Type _) [Top α] [Top β] extends
+  FunLike F α fun _ => β where
+  /-- A `TopHomClass` morphism preserves the top element. -/
+  map_top (f : F) : f ⊤ = ⊤
+#align top_hom_class TopHomClass
+
+/-- `bot_hom_class F α β` states that `F` is a type of `⊥`-preserving morphisms.
+
+You should extend this class when you extend `bot_hom`. -/
+class BotHomClass (F : Type _) (α β : outParam <| Type _) [Bot α] [Bot β] extends
+  FunLike F α fun _ => β where
+  /-- A `BotHomClass` morphism preserves the bottom element. -/
+  map_bot (f : F) : f ⊥ = ⊥
+#align bot_hom_class BotHomClass
+
+/-- `bounded_order_hom_class F α β` states that `F` is a type of bounded order morphisms.
+
+You should extend this class when you extend `bounded_order_hom`. -/
+class BoundedOrderHomClass (F : Type _) (α β : outParam <| Type _) [LE α] [LE β] [BoundedOrder α]
+  [BoundedOrder β] extends RelHomClass F ((· ≤ ·) : α → α → Prop) ((· ≤ ·) : β → β → Prop) where
+  /-- Morphisms preserve the top element. The preferred spelling is `_root_.map_top`. -/
+  map_top (f : F) : f ⊤ = ⊤
+  /-- Morphisms preserve the bottom element. The preferred spelling is `_root_.map_bot`. -/
+  map_bot (f : F) : f ⊥ = ⊥
+#align bounded_order_hom_class BoundedOrderHomClass
+
+end
+
+export TopHomClass (map_top)
+
+export BotHomClass (map_bot)
+
+attribute [simp] map_top map_bot
+
+-- Porting note: the `BoundedOrder` parameters can't be inferred through unification in all cases
+-- so they should really be instance parameters once this is technically possible.
+-- We have to supply these instances through `letI` in some cases as a work-around.
+-- See note [lower instance priority]
+instance (priority := 100) BoundedOrderHomClass.toTopHomClass {_ : LE α} {_ : LE β}
+    {_ : BoundedOrder α} {_ : BoundedOrder β} [BoundedOrderHomClass F α β] : TopHomClass F α β :=
+  { ‹BoundedOrderHomClass F α β› with }
+#align bounded_order_hom_class.to_top_hom_class BoundedOrderHomClass.toTopHomClass
+
+-- See note [lower instance priority]
+instance (priority := 100) BoundedOrderHomClass.toBotHomClass {_ : LE α} {_ : LE β}
+    {_ : BoundedOrder α} {_ : BoundedOrder β} [BoundedOrderHomClass F α β] : BotHomClass F α β :=
+  { ‹BoundedOrderHomClass F α β› with }
+#align bounded_order_hom_class.to_bot_hom_class BoundedOrderHomClass.toBotHomClass
+
+@[nolint dangerousInstance] -- The `OrderTop`s should be instance parameters but depend on outParams
```

Comment:
This lint *is* accurate in that this instance can fail to be inferred sometimes, see the `letI`s in `map_eq_top_iff`. Still, there doesn't seem to be another way to make this instance work. See [Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/287929-mathlib4/topic/Coercion.20to.20function.20dificulty).